### PR TITLE
Third try at adding no-GIL support

### LINKED ISSUE #2303: Free-threaded support?
Has anyone here given thought on testing this module with the no-GIL version of the Python interpreter yet?

### PR DESCRIPTION
**Pull request by clin1234**

### CODE REVIEW

**Comment by clin1234** [comment]
I'm not sure which, but reverting lambda changes in a file causes the free-threaded interpreters to fail in tests

**Comment by Avasam** [comment]
This is looking great now !

> I'm not sure which, but reverting lambda changes in a file causes the free-threaded interpreters to fail in tests

Are you sure? That would seem like a pretty bad CPython edge case bug if that was the case. Given it's truly only formatting changes

In https://github.com/Avasam/pywin32/pull/29 I reverted them all and the builds + tests are still passing.

Maybe you just hit https://github.com/mhammond/pywin32/issues/2203

I don't mind the changes either way. I'm just pretty sure they're unnecessary.

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -100,7 +101,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        # pythonarm64 NuGet has no download for Python ~=3.9.11
+        # pythonarm64 NuGet's has no download for Python ~=3.9.11
```
Comment (mhammond) [inline .github/workflows/main.yml]: this change is wrong

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -35,6 +35,7 @@ jobs:
     env:
       # TODO: We can't yet run tests with PYTHONDEVMODE=1, let's emulated it as much as we can
       # https://docs.python.org/3/library/devmode.html#effects-of-the-python-development-mode
+      # PYTHONVERBOSE: 1
```
Comment (mhammond) [inline .github/workflows/main.yml]: why this?

File: com/win32com/server/register.py [hunk 3]
```diff
@@ -103,9 +103,9 @@ def _find_localserver_exe(mustfind):
     if sys.platform != "win32":
         return sys.executable
     if os.path.splitext(os.path.basename(pythoncom.__file__))[0].endswith("_d"):
-        exeBaseName = "pythonw_d.exe"
```
Comment (mhammond) [inline com/win32com/server/register.py:106]: why this?

File: com/win32com/src/univgw.cpp [hunk 4]
```diff
@@ -722,8 +722,11 @@ BOOL initunivgw(PyObject *parentDict)
     if (!module) /* Eeek - some serious error! */
         return FALSE;
 
-    //	PyObject *dict = PyModule_GetDict(module);
-    //	if (!dict) return; /* Another serious error!*/
+        //	PyObject *dict = PyModule_GetDict(module);
```
Comment (mhammond) [inline com/win32com/src/univgw.cpp:725]: you've change indentation for no reason here - what you added here should probably be in the caller of this.

File: com/win32com/test/testPyComTest.py [hunk 5]
```diff
@@ -31,6 +31,7 @@
 win32com.test.util.RegisterPythonServer(
     os.path.join(os.path.dirname(__file__), "..", "servers", "test_pycomtest.py"),
     "Python.Test.PyCOMTest",
+    True,
```
Comment (mhammond) [inline com/win32com/test/testPyComTest.py]: not necessary?

File: pythonwin/Win32uiHostGlue.h [hunk 6]
```diff
@@ -145,6 +145,8 @@ inline BOOL Win32uiHostGlue::DynamicApplicationInit(const TCHAR *cmd, const TCHA
     TCHAR py_dll[20];
 #ifdef _DEBUG
     wsprintf(py_dll, _T("Python%d%d_d.dll"), PY_MAJOR_VERSION, PY_MINOR_VERSION);
+#elif defined(Py_GIL_DISABLED)
```
Comment (mhammond) [inline pythonwin/Win32uiHostGlue.h:152]: is there a `t_d` like story here? That wouldn't block this though

File: win32/src/PyWinTypes.h [hunk 7]
```diff
@@ -14,7 +14,7 @@
 #undef small
 #endif
 
-#include "Python.h"
+#include <Python.h>
```
Comment (mhammond) [inline win32/src/PyWinTypes.h]: why this?

File: win32/src/PyWinTypes.h [hunk 8]
```diff
@@ -38,16 +38,30 @@
...
 // To setup the module object itself and the module's dictionary.
+#ifdef Py_GIL_DISABLED
 #define PYWIN_MODULE_INIT_PREPARE(module_name, functions, docstring)                                        \
     PyObject *dict, *module;                                                                                \
     static PyModuleDef module_name##_def = {PyModuleDef_HEAD_INIT, #module_name, docstring, -1, functions}; \
     if (PyWinGlobals_Ensure() == -1)                                                                        \
         return NULL;                                                                                        \
     if (!(module = PyModule_Create(&module_name##_def)))                                                    \
         return NULL;                                                                                        \
+    PyUnstable_Module_SetGIL(module, Py_MOD_GIL_NOT_USED);                                                  \
```
Comment (mhammond) [inline win32/src/PyWinTypes.h:49]: it would be lovely to not duplicate all this for adding one line, but macros do suck, so whatever :)

**Review by mhammond — CHANGES_REQUESTED** [review]
This is great, thanks. Lots of unrelated stuff I'd like removed though.

This enables pywin32 to build in this nogil environment, but not be safe there. But it's a great first step towards that.

I'm fine with this if we:
* don't have CI publish artifacts. Tests are fine, wheels aren't :)
* top-level README note about the nogil status, explaining no artifacts, pointing at...
* new NOGIL.md or whatever, listing a longer status/todo before we can release artifacts. eg, the obvious-but-hidden stuff claude pointed out. claude would be good building that later, but for now a skeleton based on that other transcript you posted would be fine.

File: pythonwin/Win32uiHostGlue.h (hunk 6 above)
Comment (clin1234) [reply pythonwin/Win32uiHostGlue.h:152]: Apparently it's `td`: https://github.com/mhammond/pywin32/pull/2686#discussion_r3150533245

File: com/win32com/src/univgw.cpp (hunk 4 above)
Comment (clin1234) [reply com/win32com/src/univgw.cpp:725]: I only got CI to shut up about clang-format by checking out e390f52965177bd630a2773dcc3cb562a50fe2aa

File: com/win32com/server/register.py (hunk 3 above)
Comment (clin1234) [reply com/win32com/server/register.py:106]: Reverting back to the _d exes causes CI to fail: https://github.com/mhammond/pywin32/actions/runs/27096523197?pr=2767

**Review by mhammond — APPROVED** [review]
This is fantastic, thanks so much!

**Comment by mscheltienne** [comment]
🚀
